### PR TITLE
prov/efa: move DATA packet related functions to a separate file.

### PR DESCRIPTION
**Pull request by wzamazon**
This patch move functions related to DATA packet related funcitons
to a separate file: rxr_pkt_type_data.c

Specifically,

the following functions from rxr_ep.c was moved and renamed:

rxr_ep_send_data_pkt_entry() -> rxr_pkt_send_data()
rxr_copy_from_iov() -> rxr_copy_from_iov()
rxr_ep_mr_send_data_pkt_entry() -> rxr_pkt_send_data_mr_cache()

the following functions from rxr_cq.c was moved and renamed:

rxr_cq_handle_data() -> rxr_pkt_handle_data_recv()
rxr_cq_handle_pkt_with_data() -> rxr_pkt_proc_data()

Signed-off-by: Wei Zhang <wzam@amazon.com>

### CODE REVIEW

**Comment by wzamazon** [comment]
@ofiwg/aws-efa-team please review!

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
the failure is due connection time out:

 timeout 5m /home/ec2-user/anaconda3/bin/mpirun -n 2 -N 1 -x 'FI_PROVIDER=tcp;ofi_rxm' -x FI_EFA_ENABLE_SHM_TRANSFER=0 -x LD_LIBRARY_PATH=/home/ec2-user/anaconda3/lib/:/home/ec2-user/aws-ofi-nccl/install/lib/:/home/ec2-user/nccl/build/lib:/usr/local/cuda-10.1:/home/ec2-user/libfabric/install/lib/:/home/ec2-user/rdma-core/build/lib/:/home/ec2-user/libfabric/install/lib/:/home/ec2-user/rdma-core/build/lib/:/usr/lib64/openmpi/lib/:/usr/local/cuda/lib64:/usr/local/lib:/usr/lib:/usr/local/cuda/extras/CUPTI/lib64:/usr/local/mpi/lib:/lib/:/usr/lib64/openmpi/lib/:/usr/local/cuda/lib64:/usr/local/lib:/usr/lib:/usr/local/cuda/extras/CUPTI/lib64:/usr/local/mpi/lib:/lib/:/lib:/usr/local/cuda/lib64:/usr/local/lib:/usr/lib:/usr/local/cuda/extras/CUPTI/lib64:/usr/local/cuda/efa/lib:/usr/local/cuda/lib:/opt/amazon/efa/lib64:/usr/local/mpi/lib: --mca btl tcp,self --mca btl_tcp_if_exclude lo,docker0 --bind-to none --tag-output --hostfile hosts /home/ec2-user/aws-ofi-nccl/install/bin/nccl_connection
Warning: Permanently added '172.31.0.164' (ECDSA) to the list of known hosts.
ssh: connect to host 172.31.0.220 port 22: Connection timed out
--------------------------------------------------------------------------

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
@ofiwg/aws-efa-team  Added two commits to this PR, please review again. Thank you very much!

**Comment by wzamazon** [comment]
@shefty  Would you please share the error message of "continuous-integration/jenkins/pr-head"?

**Comment by shefty** [comment]
```
stderr: fatal: unable to access 'https://github.com/ofiwg/libfabric.git/': TCP connection reset by peer
```
Some connection problem.  You can ignore

File: prov/efa/src/rxr/rxr_pkt_type_misc.c [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+
+	rxr_pkt_entry_release_rx(ep, pkt_entry);
+
+	if (tx_entry->state != RXR_TX_SEND) {
+		tx_entry->state = RXR_TX_SEND;
+		dlist_insert_tail(&tx_entry->entry, &ep->tx_pending_list);
+	}
+}
+
+/*  READRSP pakcet functions */
```
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_pkt_type_misc.c]: Typo here

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c]: Thank you! fixed in new revision.

File: prov/efa/src/rxr/rxr_pkt_cmd.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _RXR_PKT_CMD_H
+#define _RXR_PKT_CMD_H
+
+struct rxr_ep;
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_cmd.h]: rxr.h include this file at the beginning, therefore this file cannot include rxr.h

File: prov/efa/src/rxr/rxr_pkt_cmd.c [hunk 3]
```diff
@@ -0,0 +1,618 @@
...
+		break;
+	}
+
+	return ret;
+}
+
+/*
+ *   rxr_pkt_handle_ctrl_sent() uses handle_sent() functions declared in rxr_pkt_type.h
+ */
+void rxr_pkt_handle_ctrl_sent(struct rxr_ep *rxr_ep, struct rxr_pkt_entry *pkt_entry)
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_cmd.c]: Yes. changed in new revision

File: prov/efa/src/rxr/rxr_pkt_cmd.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (shefty) [reply prov/efa/src/rxr/rxr_pkt_cmd.h]: That this header file doesn't have any includes in worrisome.  There are several structures that need to be defined before this file is included, forcing an include order.  Heck, this file doesn't even include what it needs for C data types.

Given that it's sole purpose is to incompletely declare 5 functions (6 with debug), is it even needed?

[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_cmd.h]: Good point! I push a new revision, in which I made the following changes:

1. rxr_pkt_cmd.h now includes "rxr.h" and remove "struct rxr_ep" etc.
2. rxr.h does not include "rxr_pkt_cmd.h" any more.
3. C files that used functions in rxr_pkt_cmd.h now include it.

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
the aws CI failure is capacity issue

**Pull request merged**